contracts/multiswap-contracts/FiberRouter.sol

```diff
@@ -228,116 +237,71 @@ contract FiberRouter is Ownable, TokenReceivable {
-     * @param _cctpTokenMessenger The address of the CCTP Token Messenger contract.
-     * @param _usdcToken The address of the USDC token contract.
-     * @param _sourceCCTPFundManager The address of the fund manager on the source network.
-     **/
-    function initCCTP(
-        address _cctpTokenMessenger,
-        address _usdcToken,
-        address _sourceCCTPFundManager
-    ) external onlyOwner {
-        require(_cctpTokenMessenger != address(0), "FR: Invalid CCTP Token Messenger address");
-        require(_usdcToken != address(0), "FR: Invalid USDC Token address");
-        require(_sourceCCTPFundManager != address(0), "FR: Invalid Source CCTP Fund Manager address");
-
-        cctpTokenMessenger = _cctpTokenMessenger;
-        usdcToken = _usdcToken;
-        sourceCCTPFundManager = _sourceCCTPFundManager;
-    }
 
     /**
      * @dev Initiates a token swap.
      * @param token The token to be swapped.
      * @param amount The amount to be swapped.
-     * @param targetNetwork The target network for the swap.
-     * @param targetToken The target token for the swap.
-     * @param targetAddress The target address for the swap.
      * @param withdrawalData Data related to the withdrawal.
      * @param cctpType Boolean indicating whether it's a CCTP swap.
      */
-    function swap(
+    function swapSigned(
         address token,
         uint256 amount,
-        uint256 targetNetwork,
-        address targetToken,
-        address targetAddress,
+        SwapCrossData memory sd,
         bytes32 withdrawalData,
-        bool cctpType
+        bool cctpType,
+        FeeDistributionData memory fd
     ) external payable nonReentrant {
         // Validation checks
         require(token != address(0), "FR: Token address cannot be zero");
-        require(targetToken != address(0), "FR: Target token address cannot be zero");
-        require(targetNetwork != 0, "FR: Target network is required");
-        require(targetAddress != address(0), "FR: Target address cannot be zero");
+        require(sd.targetToken != address(0), "FR: Target token address cannot be zero");
+        require(sd.targetNetwork != 0, "FR: Target network is required");
+        require(sd.targetAddress != address(0), "FR: Target address cannot be zero");
         require(amount != 0, "FR: Amount must be greater than zero");
         require(withdrawalData != 0, "FR: Withdraw data cannot be empty");
         require(msg.value != 0, "FR: Gas Amount must be greater than zero");
 
-        // Transfer the gas fee to the gasWallet
-        (bool success, ) = payable(gasWallet).call{value: msg.value}("");
-        require(success, "FR: Gas fee transfer failed");
+        // Transfer tokens to FiberRouter
+        amount = SafeAmount.safeTransferFrom(token, _msgSender(), address(this), amount);
+
+        // Distribute the fees
+        uint256 amountOut = _distributeFees(token, amount, fd);
```

Comment:
This is be quiet expensive to be done on every swap. Involves multiple transfers and signature check, etc.

Fine for now but lets have a task to re-design the fee mechanism to be more efficient. For example, if there are repeatable patterns, we can have a list of registered patterns, and assign the fee amount to the specific pattern. And later distribute in aggregate.

Reply:
According to discussion on slack, it's fine to eat this efficiency cost for the sake of flexibility for the first version.